File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/ContentEntryStatus.java
package com.ustadmobile.lib.db.entities;

import com.ustadmobile.lib.database.annotation.UmEntity;
import com.ustadmobile.lib.database.annotation.UmPrimaryKey;

@UmEntity
public class ContentEntryStatus {


    public static final int LOCAL_STATUS_UNAVAILABLE = 0;

    public static final int LOCAL_STATUS_AVAILABLE = 1;

    //Always equals contentEntryUid
    @UmPrimaryKey
    private long cesUid;

    private long totalSize;

    private long bytesDownloadSoFar;

    private int downloadStatus;

    private boolean locallyAvailable;

    private int downloadSpeed;

    private boolean invalidated = false;

    private boolean cesLeaf;

    public ContentEntryStatus() {

    }

    public ContentEntryStatus(long contentEntryUid, boolean isLeaf, long totalSize) {
        this.cesUid = contentEntryUid;
        this.cesLeaf = isLeaf;
        this.totalSize = totalSize;
    }

    public long getCesUid() {
        return cesUid;
    }

    public void setCesUid(long cesUid) {
        this.cesUid = cesUid;
    }

    public boolean isInvalidated() {
        return invalidated;
    }

    public void setInvalidated(boolean invalidated) {
        this.invalidated = invalidated;
    }

    public boolean isCesLeaf() {
        return cesLeaf;
    }

    public void setCesLeaf(boolean cesLeaf) {
        this.cesLeaf = cesLeaf;
    }

    public long getTotalSize() {
        return totalSize;
    }

    public void setTotalSize(long totalSize) {
        this.totalSize = totalSize;
    }

    public long getBytesDownloadSoFar() {
        return bytesDownloadSoFar;
    }

    public void setBytesDownloadSoFar(long bytesDownloadSoFar) {
        this.bytesDownloadSoFar = bytesDownloadSoFar;
    }

    public int getDownloadStatus() {
        return downloadStatus;
    }

    public void setDownloadStatus(int downloadStatus) {
        this.downloadStatus = downloadStatus;
    }


    public boolean isLocallyAvailable() {
        return locallyAvailable;
    }

    public void setLocallyAvailable(boolean locallyAvailable) {
        this.locallyAvailable = locallyAvailable;
    }

    public int getDownloadSpeed() {
        return downloadSpeed;
    }

    public void setDownloadSpeed(int downloadSpeed) {
        this.downloadSpeed = downloadSpeed;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ContentEntryStatus that = (ContentEntryStatus) o;

        if (cesUid != that.cesUid) return false;
        if (totalSize != that.totalSize) return false;
        if (bytesDownloadSoFar != that.bytesDownloadSoFar) return false;
        if (downloadStatus != that.downloadStatus) return false;
        if (invalidated != that.invalidated) return false;
        return cesLeaf == that.cesLeaf;
    }

    @Override
    public int hashCode() {
        int result = (int) (cesUid ^ (cesUid >>> 32));
        result = 31 * result + (int) (totalSize ^ (totalSize >>> 32));
        result = 31 * result + (int) (bytesDownloadSoFar ^ (bytesDownloadSoFar >>> 32));
        result = 31 * result + downloadStatus;
        result = 31 * result + (invalidated ? 1 : 0);
        result = 31 * result + (cesLeaf ? 1 : 0);
        return result;
    }
}

File: lib-database-entities/src/main/java/com/ustadmobile/lib/db/entities/ContentEntryStatus.kt
package com.ustadmobile.lib.db.entities

import com.ustadmobile.lib.database.annotation.UmEntity
import com.ustadmobile.lib.database.annotation.UmPrimaryKey

@UmEntity
class ContentEntryStatus {

    //Always equals contentEntryUid
    @UmPrimaryKey
    var cesUid: Long = 0

    var totalSize: Long = 0

    var bytesDownloadSoFar: Long = 0

    var downloadStatus: Int = 0

    var isLocallyAvailable: Boolean = false

    var downloadSpeed: Int = 0

    var isInvalidated = false

    var isCesLeaf: Boolean = false

    constructor()

    constructor(contentEntryUid: Long, isLeaf: Boolean, totalSize: Long) {
        this.cesUid = contentEntryUid
        this.isCesLeaf = isLeaf
        this.totalSize = totalSize
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false

        val that = o as ContentEntryStatus?

        if (cesUid != that!!.cesUid) return false
        if (totalSize != that.totalSize) return false
        if (bytesDownloadSoFar != that.bytesDownloadSoFar) return false
        if (downloadStatus != that.downloadStatus) return false
        return if (isInvalidated != that.isInvalidated) false else isCesLeaf == that.isCesLeaf
    }

    override fun hashCode(): Int {
        var result = (cesUid xor cesUid.ushr(32)).toInt()
        result = 31 * result + (totalSize xor totalSize.ushr(32)).toInt()
        result = 31 * result + (bytesDownloadSoFar xor bytesDownloadSoFar.ushr(32)).toInt()
        result = 31 * result + downloadStatus
        result = 31 * result + if (isInvalidated) 1 else 0
        result = 31 * result + if (isCesLeaf) 1 else 0
        return result
    }

    companion object {


        const val LOCAL_STATUS_UNAVAILABLE = 0

        const val LOCAL_STATUS_AVAILABLE = 1
    }
}
